Mobile backward compatibility

iseeyuan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#42413 Mobile backward compatibility**

When a default argument is added, it does not break backward compatibility (BC) for full-jit, but does break BC for mobile bytecode. For example, https://github.com/pytorch/pytorch/pull/40737. To make bytecode BC in this case, we

1. Introduce kMinSupportedBytecodeVersion. The loaded model version should be between kMinSupportedBytecodeVersion and kProducedBytecodeVersion.
2. If an operator is updated, and we can handle BC, bump the kProducedBytecodeVersion (for example, from 3 to 4).
3. If model version is at the older version of the operator, add an adapter function at loading. For the added default arg, we push this default arg to stack before calling the actual operator function.

Differential Revision: [D22898314](https://our.internmc.facebook.com/intern/diff/D22898314)

iseeyuan committed:
Mobile backward compatibility

[ghstack-poisoned]

iseeyuan committed:
Update on "Mobile backward compatibility"




[ghstack-poisoned]

iseeyuan committed:
Update on "Mobile backward compatibility"




[ghstack-poisoned]

iseeyuan committed:
Update on "Mobile backward compatibility"


When a default argument is added, it does not break backward compatibility (BC) for full-jit, but does break BC for mobile bytecode. For example, https://github.com/pytorch/pytorch/pull/40737. To make bytecode BC in this case, we

1. Introduce kMinSupportedBytecodeVersion. The loaded model version should be between kMinSupportedBytecodeVersion and kProducedBytecodeVersion.
2. If an operator is updated, and we can handle BC, bump the kProducedBytecodeVersion (for example, from 3 to 4).
3. If model version is at the older version of the operator, add an adapter function at loading. For the added default arg, we push this default arg to stack before calling the actual operator function. 


[ghstack-poisoned]

iseeyuan committed:
Update on "Mobile backward compatibility"


When a default argument is added, it does not break backward compatibility (BC) for full-jit, but does break BC for mobile bytecode. For example, https://github.com/pytorch/pytorch/pull/40737. To make bytecode BC in this case, we

1. Introduce kMinSupportedBytecodeVersion. The loaded model version should be between kMinSupportedBytecodeVersion and kProducedBytecodeVersion.
2. If an operator is updated, and we can handle BC, bump the kProducedBytecodeVersion (for example, from 3 to 4).
3. If model version is at the older version of the operator, add an adapter function at loading. For the added default arg, we push this default arg to stack before calling the actual operator function. 


[ghstack-poisoned]

mruberry commented on torch/csrc/jit/mobile/function.cpp:
```diff
@@ -49,6 +49,15 @@ bool Function::append_operator(
   }
 
+  if (model_version == 0x3L &&
+      model_version < caffe2::serialize::kProducedBytecodeVersion &&
```
Following the PR note, is the idea that the _convolution PR will bump kProducedBytecodeVersion and make this check true? 

Should there also be a kMaxSupportedBytecodeVersion, or is that redundant with kMaxSupportedFileFormatVersion?

mruberry replied:
(Note on the server side we have separate notions of produced / max supported because the produced version is dynamic based on the operations in the graph.)

iseeyuan replied:
> Following the PR note, is the idea that the _convolution PR will bump kProducedBytecodeVersion and make this check true?
Yes.
> 
> Should there also be a kMaxSupportedBytecodeVersion, or is that redundant with kMaxSupportedFileFormatVersion or kProducedBytecodeVersion?
For this use case (adding default arg to _convolution), we have the new code (version 4) supports both the old (v3) and new (v4) models. Full-jit does not need to change version numbers for this kind of update? It does hurt for old code (v3) running new models without _convolution. For this PR we don't support this.

I'm wondering if it's helpful to have dynamic version, so that the old code supports the new models without _convolution?
That's my understanding of dynamic versioning. Correct me if I'm wrong. 

Later for example if _relu is updated. When serializing bytecode, we check:
if _relu in model: 
  kProducedBytecodeVersion = 5
else if _convolution in model:
  kProducedBytecodeVersion = 4
else:
  kProducedBytecodeVersion = 3

Pro: for some model without curtain ops, it could run with bigger range of code versions.
Con: with more and more updated symbols, with logic could be complicated.

How do you think, @ljk53 @dreiss @xcheng16 ?

mruberry replied:
> I'm wondering if it's helpful to have dynamic version, so that the old code supports the new models without _convolution?
> That's my understanding of dynamic versioning. Correct me if I'm wrong.

Your understanding is correct.
 
> Later for example if _relu is updated. When serializing bytecode, we check:
> if _relu in model:
> kProducedBytecodeVersion = 5
> else if _convolution in model:
> kProducedBytecodeVersion = 4
> else:
> kProducedBytecodeVersion = 3
>

Yep.

> Pro: for some model without curtain ops, it could run with bigger range of code versions.
> Con: with more and more updated symbols, with logic could be complicated.

That's true. We decided to do it on server, but it may not be necessary for mobile.

iseeyuan replied:
As discussed with @ljk53 , we may keep it as is for now. Later some FC may be supported from active service, or we could come back and re-visit the option of dynamic versioning.

mruberry commented:
Hey @iseeyuan, I had a chance to review this and your responses more thoroughly with @ngimel. 

We have three questions:

- Should this PR support bytecode version 4 even though it produces bytecode version 3? 
  - If not, what will happen when bytecode version 4 is produced?
  - For example, do any Sandcastle builds run mobile-PyTorch services that won't know how to read version 4 bytecode?
- Is there a way to add tests for the translation behavior implemented in this PR? 
- When the _convolution PR is rebased on this, what should that PR test to validate this translation works as expected? 
  - Should it just up the produced bytecode version to 4 and let some other tests validate its behavior?
  - Should that PR produce a test fixture of a version 3 convolution and verify it can be loaded and run correctly?

iseeyuan commented:
Apology for the delayed reply because of my PTO. My answers are embedded below. Please let me know if they make sense. 

> Hey @iseeyuan, I had a chance to review this and your responses more thoroughly with @ngimel.
> 
> We have three questions:
> 
> * Should this PR support bytecode version 4 even though it produces bytecode version 3?
No. The generated model version for this PR is still version 3. In mobile/import.cpp, we check kMinSupportedBytecodeVersion <= model_version <= kProducedBytecodeVersion
>   
>   * If not, what will happen when bytecode version 4 is produced?
bytecode version 4 cannot be produced with this PR, because kProducedBytecodeVersion = 3.
>   * For example, do any Sandcastle builds run mobile-PyTorch services that won't know how to read version 4 bytecode?
> * Is there a way to add tests for the translation behavior implemented in this PR?
The translation behavior of this PR only happens with the _convolution PR + (kProducedBytecodeVersion = 4), we call it PR_1. I tested it locally. PR_1 on top of this PR works with both bytecode version 3 and version 4.
> * When the _convolution PR is rebased on this, what should that PR test to validate this translation works as expected?
Like I mentioned above. Specifically, 
1) In test_lite_interpreter, testLiteInterpreterConv covers PR_1 works with bytecode version 4. 
2) Hack testLiteInterpreterConv in the old code to generate version 3 bytecode, and load/run/test it in PR_1. 
>   
>   * Should it just up the produced bytecode version to 4 and let some other tests validate its behavior?
I think it's this option. If PR_1 is imported, there are a number of CI production models with _convolution. 
>   * Should that PR produce a test fixture of a version 3 convolution and verify it can be loaded and run correctly?
It can be done locally. If we could upload the version 3 bytecode generated from step 2) mentioned above, we could add an oss test with the model file. However, since bytecode format is currently used internal only, I doubt if it's worthwhile.

mruberry commented:
Thank you for the update @iseeyuan, that sounds reasonable. Ping @ngimel to verify this is good to go.

ngimel commented:
Thank you @iseeyuan, this looks good, we'll work with nvidia to update convolution PR to produce kProducedBytecode 4, and to also test the bc behavior. Will let you know if there are any issues.

ngimel reviewed (approved):
Approving on my side, but would be good to get approvals from mobile team also.

ljk53 commented on torch/csrc/jit/mobile/function.cpp:
```diff
@@ -49,6 +49,15 @@ bool Function::append_operator(
 
+  if (model_version == 0x3L &&
+      model_version < caffe2::serialize::kProducedBytecodeVersion &&
+      opname == c10::OperatorName("aten::_convolution", "")) {
```
We probably can consider adding a comment to point to the PR that introduces the schema change, or put the old and new schema in the comment directly.

ljk53 reviewed (approved):
LGTM

iseeyuan committed:
Update on "Mobile backward compatibility"


When a default argument is added, it does not break backward compatibility (BC) for full-jit, but does break BC for mobile bytecode. For example, https://github.com/pytorch/pytorch/pull/40737. To make bytecode BC in this case, we

1. Introduce kMinSupportedBytecodeVersion. The loaded model version should be between kMinSupportedBytecodeVersion and kProducedBytecodeVersion.
2. If an operator is updated, and we can handle BC, bump the kProducedBytecodeVersion (for example, from 3 to 4).
3. If model version is at the older version of the operator, add an adapter function at loading. For the added default arg, we push this default arg to stack before calling the actual operator function.

Differential Revision: [D22898314](https://our.internmc.facebook.com/intern/diff/D22898314)

[ghstack-poisoned]

iseeyuan committed:
Update on "Mobile backward compatibility"


When a default argument is added, it does not break backward compatibility (BC) for full-jit, but does break BC for mobile bytecode. For example, https://github.com/pytorch/pytorch/pull/40737. To make bytecode BC in this case, we

1. Introduce kMinSupportedBytecodeVersion. The loaded model version should be between kMinSupportedBytecodeVersion and kProducedBytecodeVersion.
2. If an operator is updated, and we can handle BC, bump the kProducedBytecodeVersion (for example, from 3 to 4).
3. If model version is at the older version of the operator, add an adapter function at loading. For the added default arg, we push this default arg to stack before calling the actual operator function.

Differential Revision: [D22898314](https://our.internmc.facebook.com/intern/diff/D22898314)

[ghstack-poisoned]
